--- coq.8.4.2/configure	2013-04-04 15:13:27.000000000 +0200
+++ coq.8.4.2/configure	2013-08-19 10:06:53.354192793 +0200
@@ -521,16 +521,22 @@
 	    echo "Configuration script failed!"
 	    exit 1
 	fi
-    elif [ -f "${CAMLLIB}/camlp5/${CAMLP4MOD}.cma" ]; then
-        CAMLP4LIB=+camlp5
-	FULLCAMLP4LIB=${CAMLLIB}/camlp5
-    elif [ -f "${CAMLLIB}/site-lib/${CAMLP4MOD}.cma" ]; then
+    else
+	camlp5dir="$(camlp5 -where)"
+ 	if [ "$camlp5dir" != "" ]; then
+	    CAMLP4LIB=$camlp5dir
+	    FULLCAMLP4LIB=$camlp5dir
+	elif [ -f "${CAMLLIB}/camlp5/${CAMLP4MOD}.cma" ]; then
+            CAMLP4LIB=+camlp5
+	    FULLCAMLP4LIB=${CAMLLIB}/camlp5
+	elif [ -f "${CAMLLIB}/site-lib/${CAMLP4MOD}.cma" ]; then
         CAMLP4LIB=+site-lib/camlp5
 	FULLCAMLP4LIB=${CAMLLIB}/site-lib/camlp5
-    else
-	echo "No Camlp5 installation found. Looking for Camlp4 instead..."
-	usecamlp5=no
-    fi
+	else
+	    echo "No Camlp5 installation found. Looking for Camlp4 instead..."
+	    usecamlp5=no
+	fi
+   fi
 esac
 
 # If we're (still...) going to use Camlp5, let's check its version
