diff --git a/Makefile b/Makefile
index 22f9374e..05726096 100644
--- a/Makefile
+++ b/Makefile
@@ -69,6 +69,7 @@ COMPCERT ?= platform
 ZLIST ?= bundled
 ARCH ?=
 BITSIZE ?=
+COQEXTRAFLAGS ?=
 
 # # Internal variables #
 # Set to true if the bundled CompCert is used
@@ -340,6 +341,7 @@ $(info COMPCERT_EXPLICIT_PATH=$(COMPCERT_EXPLICIT_PATH))
 $(info COMPCERT_BUILD_FROM_SRC=$(COMPCERT_BUILD_FROM_SRC))
 $(info COMPCERT_NEW=$(COMPCERT_NEW))
 $(info COQFLAGS=$(COQFLAGS))
+$(info COQEXTRAFLAGS=$(COQEXTRAFLAGS))
 $(info COMPCERT_R_FLAGS=$(COMPCERT_R_FLAGS))
 $(info =================================)
 
@@ -685,7 +687,7 @@ IRIS_INSTALL_FILES=$(sort $(IRIS_INSTALL_FILES_SRC) $(IRIS_INSTALL_FILES_VO))
 
 # This line sets COQF depending on the folder of the input file $<
 # If the folder name contains compcert, $(COMPCERT_R_FLAGS) is added, otherwise not.
-%.vo: COQF=$(if $(findstring $(COMPCERT_SRC_DIR), $(dir $<)), $(COMPCERT_R_FLAGS), $(COQFLAGS))
+%.vo: COQF=$(if $(findstring $(COMPCERT_SRC_DIR), $(dir $<)), $(COMPCERT_R_FLAGS), $(COQFLAGS)) $(COQEXTRAFLAGS)
 
 # If CompCert changes, all .vo files need to be recompiled
 %.vo: $(COMPCERT_CONFIG)
@@ -793,8 +795,8 @@ install: VST.config
 	for f in $(EXTRA_INSTALL_FILES); do install -m 0644 $$f "$(INSTALLDIR)/$$(dirname $$f)"; done
 
 build-iris: _CoqProject
-	$(COQC) $(COQFLAGS) $(PROGSDIR)/incr.v
-	for f in $(IRIS_INSTALL_FILES_SRC); do if [ "$${f##*.}" = "v" ]; then echo COQC $$f; $(COQC) $(COQFLAGS) $$f; fi; done
+	$(COQC) $(COQFLAGS) $(COQEXTRAFLAGS) $(PROGSDIR)/incr.v
+	for f in $(IRIS_INSTALL_FILES_SRC); do if [ "$${f##*.}" = "v" ]; then echo COQC $$f; $(COQC) $(COQFLAGS) $(COQEXTRAFLAGS) $$f; fi; done
 
 install-iris: VST.config
 	install -d "$(INSTALLDIR)"
@@ -938,7 +940,7 @@ memmgr:  floyd/proofauto.vo floyd/library.vo floyd/VSU.vo
 nothing: # need this target for the degenerate case of "make -tk */*.vo" in coq-action.yml
 
 assumptions.txt: veric/tcb.vo
-	$(COQC) $(COQFLAGS) veric/tcb.v > assumptions.txt
+	$(COQC) $(COQFLAGS) $(COQEXTRAFLAGS) veric/tcb.v > assumptions.txt
 	bash util/check_assumptions.sh
 
 # $(CC_TARGET): compcert/make
@@ -950,4 +952,3 @@ assumptions.txt: veric/tcb.vo
 # such problem, not sure exactly.  -- Andrew)
 include .depend
 -include .depend-concur
-
