--- configure	2016-07-18 03:14:10.000000000 -0500
+++ configure.new	2017-01-02 13:57:44.089376511 -0600
@@ -271,7 +271,7 @@ missingtools=false
 echo "Testing Coq... " | tr -d '\n'
 coq_ver=$(${COQBIN}coqc -v 2>/dev/null | sed -n -e 's/The Coq Proof Assistant, version \([^ ]*\).*$/\1/p')
 case "$coq_ver" in
-  8.5pl2)
+  8.5pl2|8.5pl3)
         echo "version $coq_ver -- good!";;
   ?.*)
         echo "version $coq_ver -- UNSUPPORTED"
