diff --git a/Makefile b/Makefile
index af069e3f..c924283b 100644
--- a/Makefile
+++ b/Makefile
@@ -205,18 +205,19 @@ compcert.ini: Makefile.config
          echo "prepro=$(CPREPRO)"; \
          echo "linker=$(CLINKER)"; \
          echo "asm=$(CASM)"; \
-	 echo "prepro_options=$(CPREPRO_OPTIONS)";\
-	 echo "asm_options=$(CASM_OPTIONS)";\
-	 echo "linker_options=$(CLINKER_OPTIONS)";\
+         echo "prepro_options=$(CPREPRO_OPTIONS)";\
+         echo "asm_options=$(CASM_OPTIONS)";\
+         echo "linker_options=$(CLINKER_OPTIONS)";\
          echo "arch=$(ARCH)"; \
          echo "model=$(MODEL)"; \
          echo "abi=$(ABI)"; \
          echo "endianness=$(ENDIANNESS)"; \
+         echo "bitsize=$(BITSIZE)"; \
          echo "system=$(SYSTEM)"; \
          echo "has_runtime_lib=$(HAS_RUNTIME_LIB)"; \
          echo "has_standard_headers=$(HAS_STANDARD_HEADERS)"; \
          echo "asm_supports_cfi=$(ASM_SUPPORTS_CFI)"; \
-	 echo "response_file_style=$(RESPONSEFILE)";) \
+         echo "response_file_style=$(RESPONSEFILE)";) \
         > compcert.ini
 
 driver/Version.ml: VERSION
@@ -253,6 +254,7 @@ ifeq ($(INSTALL_COQDEV),true)
           install -m 0644 $$d/*.vo $(DESTDIR)$(COQDEVDIR)/$$d/; \
 	done
 	install -m 0644 ./VERSION $(DESTDIR)$(COQDEVDIR)
+	install -m 0644 ./compcert.ini $(DESTDIR)$(COQDEVDIR)
 	@(echo "To use, pass the following to coq_makefile or add the following to _CoqProject:"; echo "-R $(COQDEVDIR) compcert") > $(DESTDIR)$(COQDEVDIR)/README
 endif
 
