diff --git a/theories/Base.v b/theories/Base.v
index ae22566..b89da46 100644
--- a/theories/Base.v
+++ b/theories/Base.v
@@ -1,5 +1,4 @@
-Declare ML Module "coq-unicoq.plugin".
-Declare ML Module "MetaCoqPlugin:coq-mtac2.plugin".
+Declare ML Module "coq-mtac2.plugin".
 
 (* Declare ML Module must work without the Requires to be compatible
    with async proofs. Running it before them serves as a test
