diff -crN HoTT.orig/etc/ci/install_coq_deps.sh HoTT.new/etc/ci/install_coq_deps.sh
*** HoTT.orig/etc/ci/install_coq_deps.sh	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/etc/ci/install_coq_deps.sh	2017-03-25 21:38:27.555824254 +0100
***************
*** 7,11 ****
  sudo apt-get install -q sed grep wget tar xsltproc mercurial python-lxml python-pexpect libxml2-dev libxslt1-dev
  if [ ! -z "$BUILD_COQ" -o ! -z "$UPDATE_DEP_GRAPHS" ]
  then
!     sudo apt-get install -q ocaml camlp5
  fi
--- 7,11 ----
  sudo apt-get install -q sed grep wget tar xsltproc mercurial python-lxml python-pexpect libxml2-dev libxslt1-dev
  if [ ! -z "$BUILD_COQ" -o ! -z "$UPDATE_DEP_GRAPHS" ]
  then
!     sudo apt-get install -q ocaml camlp5 ocaml-findlib
  fi
diff -crN HoTT.orig/etc/ci/install_coq.sh HoTT.new/etc/ci/install_coq.sh
*** HoTT.orig/etc/ci/install_coq.sh	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/etc/ci/install_coq.sh	2017-03-25 21:38:27.555824254 +0100
***************
*** 39,48 ****
  fi
  echo '$ ./configure '"$@"
  ./configure "$@"
! echo '$ make coqlight'
! make coqlight READABLE_ML4=1
! echo '$ sudo make install-coqlight install-devfiles'
! sudo make install-coqlight install-devfiles
  popd
  
  popd 1>/dev/null
--- 39,50 ----
  fi
  echo '$ ./configure '"$@"
  ./configure "$@"
! echo '$ make states READABLE_ML4=1 tools'
! make states tools READABLE_ML4=1
! echo '$ sudo make install-binaries + rsync plugins theories'
! touch bin/coqtop.byte bin/coqchk stm/{proof,tac,query}workertop.cma
! sudo make install-binaries
! sudo rsync -a plugins theories /usr/local/lib/coq/
  popd
  
  popd 1>/dev/null
diff -crN HoTT.orig/.git/HEAD HoTT.new/.git/HEAD
*** HoTT.orig/.git/HEAD	2017-03-25 21:38:40.223823830 +0100
--- HoTT.new/.git/HEAD	2017-03-25 21:38:27.563824254 +0100
***************
*** 1 ****
! 57389f4e2fa2f846b74c8e26c3d8a58bc00ea762
--- 1 ----
! ref: refs/heads/mz-8.6
Binary files HoTT.orig/.git/index and HoTT.new/.git/index differ
diff -crN HoTT.orig/.git/logs/HEAD HoTT.new/.git/logs/HEAD
*** HoTT.orig/.git/logs/HEAD	2017-03-25 21:38:40.223823830 +0100
--- HoTT.new/.git/logs/HEAD	2017-03-25 21:38:27.563824254 +0100
***************
*** 1,5 ****
  0000000000000000000000000000000000000000 7a6e5edc7071354b35d795348a1282ea17c63e18 Matej Kosik <matej.kosik@inria.fr> 1489737924 +0100	clone: from https://github.com/HoTT/HoTT.git
  7a6e5edc7071354b35d795348a1282ea17c63e18 9f659aaa7b8b07112624e27588bb9f074a83cd03 Matej Kosik <matej.kosik@inria.fr> 1490473493 +0100	checkout: moving from master to mz-8.6
! 9f659aaa7b8b07112624e27588bb9f074a83cd03 57389f4e2fa2f846b74c8e26c3d8a58bc00ea762 Matej Kosik <matej.kosik@inria.fr> 1490474135 +0100	checkout: moving from mz-8.6 to 57389f4
! 57389f4e2fa2f846b74c8e26c3d8a58bc00ea762 9f659aaa7b8b07112624e27588bb9f074a83cd03 Matej Kosik <matej.kosik@inria.fr> 1490474267 +0100	checkout: moving from 57389f4e2fa2f846b74c8e26c3d8a58bc00ea762 to mz-8.6
! 9f659aaa7b8b07112624e27588bb9f074a83cd03 57389f4e2fa2f846b74c8e26c3d8a58bc00ea762 Matej Kosik <matej.kosik@inria.fr> 1490474320 +0100	checkout: moving from mz-8.6 to 57389f4
--- 1,6 ----
  0000000000000000000000000000000000000000 7a6e5edc7071354b35d795348a1282ea17c63e18 Matej Kosik <matej.kosik@inria.fr> 1489737924 +0100	clone: from https://github.com/HoTT/HoTT.git
  7a6e5edc7071354b35d795348a1282ea17c63e18 9f659aaa7b8b07112624e27588bb9f074a83cd03 Matej Kosik <matej.kosik@inria.fr> 1490473493 +0100	checkout: moving from master to mz-8.6
! 9f659aaa7b8b07112624e27588bb9f074a83cd03 39faa00c1b1e6d63b0b4e88f19b2a8a6ac7c0598 Matej Kosik <matej.kosik@inria.fr> 1490474149 +0100	checkout: moving from mz-8.6 to 39faa00
! 39faa00c1b1e6d63b0b4e88f19b2a8a6ac7c0598 9f659aaa7b8b07112624e27588bb9f074a83cd03 Matej Kosik <matej.kosik@inria.fr> 1490474226 +0100	checkout: moving from 39faa00c1b1e6d63b0b4e88f19b2a8a6ac7c0598 to mz-8.6
! 9f659aaa7b8b07112624e27588bb9f074a83cd03 f23f8df3379799fdb15498f3c56bc05e9f2aedd9 Matej Kosik <matej.kosik@inria.fr> 1490474246 +0100	checkout: moving from mz-8.6 to f23f8df
! f23f8df3379799fdb15498f3c56bc05e9f2aedd9 9f659aaa7b8b07112624e27588bb9f074a83cd03 Matej Kosik <matej.kosik@inria.fr> 1490474307 +0100	checkout: moving from f23f8df3379799fdb15498f3c56bc05e9f2aedd9 to mz-8.6
Binary files HoTT.orig/.git/modules/etc/coq-scripts/index and HoTT.new/.git/modules/etc/coq-scripts/index differ
diff -crN HoTT.orig/.gitmodules HoTT.new/.gitmodules
*** HoTT.orig/.gitmodules	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/.gitmodules	2017-03-25 21:38:27.555824254 +0100
***************
*** 1,7 ****
  [submodule "coq-HoTT"]
  	path = coq-HoTT
  	url = https://github.com/coq/coq.git
! 	branch = trunk
  [submodule "etc/coq-scripts"]
  	path = etc/coq-scripts
  	url = https://github.com/JasonGross/coq-scripts.git
--- 1,7 ----
  [submodule "coq-HoTT"]
  	path = coq-HoTT
  	url = https://github.com/coq/coq.git
! 	branch = v8.6
  [submodule "etc/coq-scripts"]
  	path = etc/coq-scripts
  	url = https://github.com/JasonGross/coq-scripts.git
diff -crN HoTT.orig/theories/Modalities/Lex.v HoTT.new/theories/Modalities/Lex.v
*** HoTT.orig/theories/Modalities/Lex.v	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/theories/Modalities/Lex.v	2017-03-25 21:37:06.483826969 +0100
***************
*** 219,227 ****
                                           (hfiber@{k i} g (g x)))).
      { refine (_ oE equiv_to_O@{u a k k} O _).
        - refine (_ oE BuildEquiv _ _
!                   (O_functor_hfiber@{u a k i k k i k k k}
!                                    (* MS: Annotation can be dropped now *)
!                                    O (@pr1 A B) (g x)) _).
          unfold hfiber.
          refine (equiv_functor_sigma' 1 _). intros y; cbn.
          refine (_ oE (equiv_moveR_equiv_V _ _)).
--- 219,225 ----
                                           (hfiber@{k i} g (g x)))).
      { refine (_ oE equiv_to_O@{u a k k} O _).
        - refine (_ oE BuildEquiv _ _
!                   (O_functor_hfiber O (@pr1 A B) (g x)) _).
          unfold hfiber.
          refine (equiv_functor_sigma' 1 _). intros y; cbn.
          refine (_ oE (equiv_moveR_equiv_V _ _)).
diff -crN HoTT.orig/theories/Modalities/Nullification.v HoTT.new/theories/Modalities/Nullification.v
*** HoTT.orig/theories/Modalities/Nullification.v	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/theories/Modalities/Nullification.v	2017-03-25 21:37:06.483826969 +0100
***************
*** 78,85 ****
      refine (Localize_ind@{a i j k}
               (null_to_local_generators@{a a} (unNul O)) A B g _); intros i.
      apply (ooextendable_over_unit@{a i j a k}); intros c.
!     refine (ooextendable_postcompose@{a a j j k k k k k k}
!                                     (* MS: Can be dropped *)
                (fun (_:Unit) => B (c tt)) _ _
                (fun u => transport@{i j} B (ap c (path_unit@{a} tt u))) _).
      refine (ooextendable_islocal _ i).
--- 78,84 ----
      refine (Localize_ind@{a i j k}
               (null_to_local_generators@{a a} (unNul O)) A B g _); intros i.
      apply (ooextendable_over_unit@{a i j a k}); intros c.
!     refine (ooextendable_postcompose
                (fun (_:Unit) => B (c tt)) _ _
                (fun u => transport@{i j} B (ap c (path_unit@{a} tt u))) _).
      refine (ooextendable_islocal _ i).
diff -crN HoTT.orig/theories/Modalities/ReflectiveSubuniverse.v HoTT.new/theories/Modalities/ReflectiveSubuniverse.v
*** HoTT.orig/theories/Modalities/ReflectiveSubuniverse.v	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/theories/Modalities/ReflectiveSubuniverse.v	2017-03-25 21:34:44.639831720 +0100
***************
*** 827,833 ****
          { refine (O_indpaths _ _ _); simpl.
            intro v; exact v. }
          exact (p u).
!       - hnf.
          rewrite O_indpaths_beta; reflexivity.
      Qed.
      Global Existing Instance inO_paths.
--- 827,833 ----
          { refine (O_indpaths _ _ _); simpl.
            intro v; exact v. }
          exact (p u).
!       - simpl.
          rewrite O_indpaths_beta; reflexivity.
      Qed.
      Global Existing Instance inO_paths.
diff -crN HoTT.orig/.travis.yml HoTT.new/.travis.yml
*** HoTT.orig/.travis.yml	2017-03-25 21:38:40.219823830 +0100
--- HoTT.new/.travis.yml	2017-03-25 21:38:27.555824254 +0100
***************
*** 20,26 ****
     - UPDATE_HTML=""       WITH_AUTORECONF=""     UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION=""
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ=""     UPDATE_OPAM=""       FORCE_COQ_VERSION=""
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION=""
!    - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="v8.5"
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="trunk"
  
   global:
--- 20,26 ----
     - UPDATE_HTML=""       WITH_AUTORECONF=""     UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION=""
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ=""     UPDATE_OPAM=""       FORCE_COQ_VERSION=""
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION=""
!    - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="v8.6"
     - UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="trunk"
  
   global:
***************
*** 31,38 ****
  matrix:
    fast_finish: true
    allow_failures:
      - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ=""     UPDATE_OPAM=""       FORCE_COQ_VERSION=""
-     - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="v8.5"
      - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="trunk"
  
  
--- 31,38 ----
  matrix:
    fast_finish: true
    allow_failures:
+     - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS="yes" BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION=""
      - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ=""     UPDATE_OPAM=""       FORCE_COQ_VERSION=""
      - env: UPDATE_HTML=""       WITH_AUTORECONF="yes"  UPDATE_QUICK_DOC=""     UPDATE_AUTOGEN=""     UPDATE_DEP_GRAPHS=""    BUILD_COQ="yes"  UPDATE_OPAM=""       FORCE_COQ_VERSION="trunk"
  
  
