diff --git a/Make b/Make
index fb7b449..83ee20d 100644
--- a/Make
+++ b/Make
@@ -1,12 +1,6 @@
 -I src
 -I tests
 -R theories Ergo
--I ../Containers/src
--R ../Containers/theories Containers
--I ../Nfix/src
--R ../Nfix/theories Nfix
--I ../Counting/src
--R ../Counting/theories Counting
 -custom "$(CAMLOPTLINK) $(ZDEBUG) $(ZFLAGS) -shared -o src/ergo_plugin.cmxs src/arith.cmx src/ergo.cmx src/ergo_plugin_mod.cmx" "src/arith.cmx src/ergo.cmx src/ergo_plugin_mod.cmx" src/ergo_plugin.cmxs
 -custom "$(CAMLLINK) -g -a -o src/ergo_plugin.cma src/arith.cmo src/ergo.cmo src/ergo_plugin_mod.cmo" "src/arith.cmo src/ergo.cmo src/ergo_plugin_mod.cmo" src/ergo_plugin.cma
 -custom "$(COQBIN)coqc $(COQDEBUG) $(COQFLAGS) $*" "%.v $(ERGO_PLUGINOPT) $(ERGO_PLUGIN)" "%.vo %.glob"
